Userland/Libraries/LibIMAP/Parser.cpp

```diff
@@ -0,0 +1,836 @@
+/*
+ * Copyright (c) 2021, Kyle Pereira <hey@xylepereira.me>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Parser.h"
+#include <AK/CharacterTypes.h>
+
+namespace IMAP {
+
+ParseStatus Parser::parse(ByteBuffer&& buffer, bool expecting_tag)
+{
+    if (m_incomplete) {
+        m_buffer += buffer;
+        m_incomplete = false;
+    } else {
+        m_buffer = move(buffer);
+        position = 0;
+        m_response = SolidResponse();
+    }
+
+    if (match("+")) {
+        consume(" ");
+        auto data = parse_while(+[](u8 x) { return x != '\r'; });
+        consume("\r\n");
+        return { true, { ContinueRequest { data } } };
+    }
+
+    while (match("*")) {
+        parse_untagged();
+    }
+
+    if (expecting_tag) {
+        if (end()) {
```

Comment:
`is_end()` or `at_end()`, this seems like you're trying to end a stream or something